File: 145-binary-tree-postorder-traversal/binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void postorder(TreeNode* root, vector<int> &nodesOrder) {
        if (! root) {
            return;
        }
        postorder(root->left, nodesOrder);
        postorder(root->right, nodesOrder);
        nodesOrder.push_back(root->val);
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> nodesOrder;
        postorder(root, nodesOrder);
        return nodesOrder;
    }
};

File: 145-binary-tree-postorder-traversal/binary-tree-postorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func postorderTraversal(root *TreeNode) []int {
    var nodesOrder []int
    postorder(root, &nodesOrder)
    return nodesOrder
}

func postorder(root *TreeNode, nodesOrder *[]int) {
    if root == nil {
        return
    }
    postorder(root.Left, nodesOrder)
    postorder(root.Right, nodesOrder)
    *nodesOrder = append(*nodesOrder, root.Val)
}
